addon/-private/system/relationships/state/relationship.js

```diff
@@ -266,6 +269,42 @@ export default class Relationship {
     this.flushCanonicalLater();
   }
 
+  /*
+    Call this method once a record deletion has been persisted
+    to purge it from both current and canonical state of all
+    relationships.
+
+    @method removeDeletedInternalModelFromInverse
+    @private
+   */
+  removeDeletedInternalModelFromInverse() {
+    if (!this.inverseKey) { return; }
+
+    // we actually want a union of members and canonicalMembers
+    // they should be disjoint but currently are not due to a bug
+    let seen = Object.create(null);
+    const internalModel = this.internalModel;
+
+    const unload = inverseInternalModel => {
+      const id = guidFor(inverseInternalModel);
+
+      if (seen[id] === undefined) {
+        const relationship = inverseInternalModel._relationships.get(this.inverseKey);
+        relationship.removeDeletedInternalModelFromOwn(internalModel);
+        seen[id] = true;
+      }
+    };
+
+    this.members.forEach(unload);
+    this.canonicalMembers.forEach(unload);
+  }
+
+  removeDeletedInternalModelFromOwn(internalModel) {
```

Comment:
The divergence (but feel free to confirm):

* `removeInternalModelFromOwn`:  is removing from current state (awaiting future canonical flush)
* `removeDeletedInternalModelFromOwn`: is removing from both current state (`members`) and canonical state (`canonicalMembers`)

The naming of these methods sorta blows, but naming is hard. So at the very least we may need to document them somewhat.

Some ideas:

```js
Removes a given internal model from the relationships currentState, but not from that relationships canonicalState. This results in an optimistic delete, and subsequent confirmation (typically server side) will update the canonicalState.

@method removeInternalModelFromOwn
@param internalModel
```


```js
Removes a given internal model from both a relationships currentState and canonical state. This typically occurs when an internalRecord's deletion is persisted.

@method removeDeletedInternalModelFromOwn
@param internalModel
```